### PR TITLE
[WIP] Core/Creature: Trainer

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:
-  
-  

## Issues Addressed:

- Closes 

## SOURCE:

TC
## Tests Performed:

- 
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
https://github.com/azerothcore/azerothcore-wotlk/issues/5420 ?

**Comment by IntelligentQuantum** [comment]
> #5420 ?

Hmmm, I had not seen this issue.

**Comment by Kitzunu** [comment]
I have tried doing this 3 times and gave up cuz of it being so big xd

**Comment by acidmanifesto** [comment]
i may regret asking this, how important is this needing to be done?

**Comment by Kitzunu** [comment]
> 

That's what I concluded too haha

**Comment by Kitzunu** [comment]
> i may regret asking this, how important is this needing to be done?

Not life or death, but it is a nice system

**Comment by acidmanifesto** [comment]
> > i may regret asking this, how important is this needing to be done?
> 
> Not life or death, but it is a nice system

well then... hold my beer, lets see what i can do the next few days

**Comment by Kitzunu** [comment]
Famous last words

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -1087,52 +1078,13 @@ bool Creature::isCanInteractWithBattleMaster(Player* player, bool msg) const
...
-}
-
-bool Creature::IsValidTrainerForPlayer(Player* player, uint32* npcFlags /*= nullptr*/) const
+bool Creature::CanResetTalents(Player* player) const
 {
-    if (!IsTrainer())
-    {
+    Trainer::Trainer const* trainer = sObjectMgr->GetTrainer(GetEntry());
+    if (!trainer)
         return false;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Creature.cpp:1085]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            trainerListSpell.ReqSkillRank = trainerSpell.ReqSkillRank;
+            std::copy(trainerSpell.ReqAbility.begin(), trainerSpell.ReqAbility.end(), trainerListSpell.ReqAbility.begin());
+        }
+
+        player->SendDirectMessage(trainerList.Write());
+    }
+
+    void Trainer::TeachSpell(Creature const* npc, Player* player, uint32 spellId) const
+    {
+        if (!IsTrainerValidForPlayer(player))
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:77]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        }
+
+        player->ModifyMoney(-moneyCost);
+
+        npc->SendPlaySpellVisual(179);
+        npc->SendPlaySpellImpact(player->GetGUID(), 362);
+
+        // learn explicitly or cast explicitly
+        if (trainerSpell->IsCastable())
+            player->CastSpell(player, trainerSpell->SpellId, true);
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:108]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        else
+            player->learnSpell(trainerSpell->SpellId, false);
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:110]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+    Spell const* Trainer::GetSpell(uint32 spellId) const
+    {
+        auto itr = std::find_if(_spells.begin(), _spells.end(), [spellId](Spell const& trainerSpell)
+        {
+            return trainerSpell.SpellId == spellId;
+        });
+
+        if (itr != _spells.end())
+            return &(*itr);
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:123]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+        return nullptr;
+    }
+
+    bool Trainer::CanTeachSpell(Player const* player, Spell const* trainerSpell) const
+    {
+        SpellState state = GetSpellState(player, trainerSpell);
+        if (state != SpellState::Available)
+            return false;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:132]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+        SpellInfo const* trainerSpellInfo = sSpellMgr->AssertSpellInfo(trainerSpell->SpellId);
+        if (trainerSpellInfo->IsPrimaryProfessionFirstRank() && !player->GetFreePrimaryProfessionPoints())
+            return false;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:136]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+        return true;
+    }
+
+    SpellState Trainer::GetSpellState(Player const* player, Spell const* trainerSpell) const
+    {
+        if (player->HasSpell(trainerSpell->SpellId))
+            return SpellState::Known;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:144]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+        // check race/class requirement
+        if (!player->IsSpellFitByClassAndRace(trainerSpell->SpellId))
+            return SpellState::Unavailable;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:148]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+        // check skill requirement
+        if (trainerSpell->ReqSkillLine && player->GetBaseSkillValue(trainerSpell->ReqSkillLine) < trainerSpell->ReqSkillRank)
+            return SpellState::Unavailable;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:152]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+        for (int32 reqAbility : trainerSpell->ReqAbility)
+            if (reqAbility && !player->HasSpell(reqAbility))
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:155]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+                return SpellState::Unavailable;
+
+        // check level requirement
+        if (player->getLevel() < trainerSpell->ReqLevel)
+            return SpellState::Unavailable;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:160]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+
+        // check race/class requirement
+        if (!player->IsSpellFitByClassAndRace(trainerSpell->SpellId))
+            return SpellState::Unavailable;
+
+        // check skill requirement
+        if (trainerSpell->ReqSkillLine && player->GetBaseSkillValue(trainerSpell->ReqSkillLine) < trainerSpell->ReqSkillRank)
+            return SpellState::Unavailable;
+
+        for (int32 reqAbility : trainerSpell->ReqAbility)
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:154]: Why int32 (not uint32)?

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+        if (player->getLevel() < trainerSpell->ReqLevel)
+            return SpellState::Unavailable;
+
+        // check ranks
+        bool hasLearnSpellEffect = false;
+        bool knowsAllLearnedSpells = true;
+        for (uint8 i = 0; i < MAX_SPELL_EFFECTS; ++i)
+        {
+            SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(trainerSpell->SpellId);
+            if (!spellInfo || spellInfo->Effects[i].Effect != SPELL_EFFECT_LEARN_SPELL)
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:168]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+                continue;
+
+            hasLearnSpellEffect = true;
+            if (!player->HasSpell(spellInfo->Effects[i].TriggerSpell))
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:172]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+                knowsAllLearnedSpells = false;
+
+            if (uint32 previousRankSpellId = sSpellMgr->GetPrevSpellInChain(spellInfo->Effects[i].TriggerSpell))
+                if (!player->HasSpell(previousRankSpellId))
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:176]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+                    return SpellState::Unavailable;
+        }
+
+        if (!hasLearnSpellEffect)
+        {
+            if (uint32 previousRankSpellId = sSpellMgr->GetPrevSpellInChain(trainerSpell->SpellId))
+                if (!player->HasSpell(previousRankSpellId))
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:183]: Trainer::GetSpellStat

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+                    return SpellState::Unavailable;
+        }
+        else if (knowsAllLearnedSpells)
+            return SpellState::Known;
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:187]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+
+        // check additional spell requirement
+        for (auto const& requirePair : sSpellMgr->GetSpellsRequiredForSpellBounds(trainerSpell->SpellId))
+            if (!player->HasSpell(requirePair.second))
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:191]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+                return SpellState::Unavailable;
+
+        return SpellState::Available;
+    }
+
+    bool Trainer::IsTrainerValidForPlayer(Player const* player) const
+    {
+        if (!GetTrainerRequirement())
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:199]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+        WorldPackets::NPC::TrainerBuySucceeded trainerBuySucceeded;
+        trainerBuySucceeded.TrainerGUID = npc->GetGUID();
+        trainerBuySucceeded.SpellID = spellId;
+        player->SendDirectMessage(trainerBuySucceeded.Write());
+    }
+
+    std::string const& Trainer::GetGreeting(LocaleConstant locale) const
+    {
+        if (_greeting[locale].empty())
+            return _greeting[DEFAULT_LOCALE];
```
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:241]: Brackets.

File: src/server/game/Entities/Creature/Trainer.cpp (hunk 13 above)
Comment (UltraNix) [inline src/server/game/Entities/Creature/Trainer.cpp:154]: Why `int32` (and not `uint32`)?

File: src/server/game/Entities/Player/PlayerGossip.cpp [hunk 22]
```diff
@@ -116,11 +116,16 @@ void Player::PrepareGossipMenu(WorldObject* source, uint32 menuId /*= 0*/, bool
...
+                {
+                    Trainer::Trainer const* trainer = sObjectMgr->GetTrainer(creature->GetEntry());
+                    if (!trainer || !trainer->IsTrainerValidForPlayer(this))
                     {
+                        LOG_ERROR("sql.sql", "GOSSIP_OPTION_TRAINER:: Player %s (GUID: %u) requested wrong gossip menu: %u at Creature: %s (Entry: %u)",
+                                     GetName().c_str(), GetGUID().GetCounter(), menu->GetGossipMenu().GetMenuId(), creature->GetName().c_str(), creature->GetEntry());
                         canTalk = false;
                     }
-                    break;
+                    // no break;
```
Comment (UltraNix) [inline src/server/game/Entities/Player/PlayerGossip.cpp:127]: `[[fallthrough]]`

File: src/server/game/Globals/ObjectMgr.cpp [hunk 23]
```diff
@@ -8295,122 +8262,184 @@ void ObjectMgr::LoadMailLevelRewards()
...
-    if (!cInfo)
-    {
-        LOG_ERROR("sql.sql", "Table `npc_trainer` contains an entry for a non-existing creature template (Entry: %u), ignoring", entry);
-        return;
-    }
+    // For reload case
+    _trainers.clear();
 
-    if (!(cInfo->npcflag & UNIT_NPC_FLAG_TRAINER))
+    std::unordered_map<int32, std::vector<Trainer::Spell>> spellsByTrainer;
```
Comment (UltraNix) [inline src/server/game/Globals/ObjectMgr.cpp:8272]: Why `int32` (and not `uint32`)?

**Review by UltraNix — CHANGES_REQUESTED** [review]
Missing brackets (mainly in `Trainer.cpp`) and `[[nodiscard]]` attribute (mainly in `Trainer.h`.

**Comment by IntelligentQuantum** [comment]
CONFLICT

**Comment by UltraNix** [comment]
> CONFLICT

FIGHT!

**Comment by IntelligentQuantum** [comment]
> > CONFLICT
> 
> FIGHT!

I CAN'T

**Pull request closed without merging**